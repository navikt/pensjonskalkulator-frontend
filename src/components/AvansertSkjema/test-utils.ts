export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
