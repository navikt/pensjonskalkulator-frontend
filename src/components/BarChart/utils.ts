import { ChartData } from './BarChart'

export const getBarChartHeight = (
  value: number,
  maxValue: number,
  maxHeight: number = 128
) => {
  if (value === 0) {
    return 0
  }
  return (value / maxValue) * maxHeight
}

export const findMaxValue = (data: ChartData[]): number => {
  return data.reduce((max, { value }) => (max > value ? max : value), 0)
}
