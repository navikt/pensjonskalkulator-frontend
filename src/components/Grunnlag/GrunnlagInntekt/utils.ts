export const validateInntektInput = (
  s: string | undefined,
  updateValidationErrorMessage: (s: string) => void
) => {
  let isValid = true
  if (s === undefined || s === '') {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.required'
    )
  } else if (isNaN(s as unknown as number) || !/^[0-9]+$/.test(s)) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.type'
    )
  } else if (parseInt(s as string, 10) > 100000000) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.max'
    )
  }
  return isValid
}
