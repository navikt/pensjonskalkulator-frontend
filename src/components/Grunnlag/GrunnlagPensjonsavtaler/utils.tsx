import { PensjonsavtaleKategori } from '@/types/enums'

type PensjonsavtaleRecord = {
  [key in PensjonsavtaleKategori]?: Pensjonsavtale[]
}

export const groupPensjonsavtalerByType = (
  pensjonsavtaler: Pensjonsavtale[]
): PensjonsavtaleRecord => {
  const record: PensjonsavtaleRecord = {}

  for (const avtale of pensjonsavtaler) {
    const avtalekategori =
      PensjonsavtaleKategori[
        avtale.kategori as keyof typeof PensjonsavtaleKategori
      ]

    if (!Array.isArray(record[avtalekategori])) {
      record[avtalekategori] = [avtale]
    } else {
      record[avtalekategori]?.push(avtale)
    }
  }
  return record
}

export function getMaanedString(maaned?: number) {
  if (maaned !== undefined && maaned > 0) {
    return ` og ${maaned} md.`
  }
  return ''
}
