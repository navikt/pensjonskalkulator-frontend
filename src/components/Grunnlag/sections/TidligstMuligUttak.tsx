import React from 'react'

import { Accordion, BodyLong, Link } from '@navikt/ds-react'

import { SectionHeader } from './components/SectionHeader'

interface Props {
  uttaksalder: Uttaksalder
}

export function TidligstMuligUttak({ uttaksalder }: Props) {
  return (
    <Accordion.Item>
      <SectionHeader
        label="Tidligst mulig uttak"
        value={
          uttaksalder.maaned === 0
            ? `${uttaksalder.aar} år`
            : `${uttaksalder.aar} år, ${uttaksalder.maaned} md.`
        }
      />
      <Accordion.Content>
        <BodyLong>
          Viser første mulige tidspunkt for uttak av 100 % alderspensjon.
          Tidspunktet kan bli endret dersom din inntekt endrer seg. Selv om du
          ikke kan ta ut 100 % alderspensjon på det tidspunktet du ønsker, kan
          det være du har mulighet til å starte uttak av en lavere uttaksgrad.
          <br />
          <br />
          Rett til AFP i privat eller offentlig sektor kan også ha betydning for
          når du tidligst kan starte uttak av pensjon. Vær oppmerksom på at
          dersom du ikke har rett til AFP, vil det kunne påvirke resultatet
          ditt.
          <br />
          <br />
          Hvorfor kan du ikke starte uttak tidligere: For å ta ut alderspensjon
          før du har fylt 67 år, må du ha tilstrekkelig høy opptjening. Ved helt
          eller delvis uttak før 67 år, må opptjeningen være så høy at pensjonen
          ved fylte 67 år minst tilsvarer
          <Link>garantipensjonsnivået</Link>. Fra fylte 67 år har alle med
          opptjente pensjonsrettigheter rett til å starte alderspensjon.
          <br />
          <br />
          Tidspunktet er et estimat og du bør sjekke igjen når tidspunktet
          nærmer seg.
        </BodyLong>
      </Accordion.Content>
    </Accordion.Item>
  )
}
