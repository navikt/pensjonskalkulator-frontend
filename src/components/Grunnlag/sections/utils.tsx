import { PensjonsavtaleKategori } from '@/types/enums'

type PensjonsavtaleRecord = {
  [key in PensjonsavtaleKategori]?: Pensjonsavtale[]
}

export const groupPensjonsavtalerByType = (
  pensjonsavtaler: Pensjonsavtale[]
): PensjonsavtaleRecord => {
  const record: PensjonsavtaleRecord = {}

  for (const avtale of pensjonsavtaler) {
    const avtalekategori =
      PensjonsavtaleKategori[
        avtale.kategori as keyof typeof PensjonsavtaleKategori
      ]

    if (!Array.isArray(record[avtalekategori])) {
      record[avtalekategori] = [avtale]
    } else {
      record[avtalekategori]?.push(avtale)
    }
  }
  return record
}

export function getPensjonsavtalerTittel(
  harSamtykket: boolean,
  showError: boolean,
  antallAvtaler: string
) {
  if (!harSamtykket) {
    return 'Ikke innhentet'
  }
  if (showError) {
    return 'Kan ikke hentes'
  }
  return antallAvtaler
}
