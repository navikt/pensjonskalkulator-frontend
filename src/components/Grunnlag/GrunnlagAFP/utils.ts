// Input:
// AFPValg
// isKap19
// erApoteker
// loependeVedtak
// uforegrad
// intl - interasjonalisering
import { IntlShape } from 'react-intl'

import { isFoedtFoer1963 } from '@/utils/alder'

interface IAfpGrunnlagInput {
  /** Dersom bruker er kap19/apoteker kan man velge å beregne offentlig AFP sammen med aldersensjon */
  afpUtregning: AfpUtregningValg
  /** Valg av hvilken AFP utregning bruker har valgt */
  afpValg: AfpRadio | null
  /** Valg for beregning av pensjon */
  beregningsvalg: Beregningsvalg | null
  /** Indikerer om brukeren er apoteker */
  erApoteker: boolean
  /** Brukerens fødselsdato */
  foedselsdato: string
  /** Informasjon om løpende vedtak */
  loependeVedtak: LoependeVedtak
  /** Samtykke til beregning av offentlig AFP */
  samtykkeOffentligAFP: boolean | null
}

export const afpContentIntl = (intl: IntlShape) => ({
  offentligIkkeBeregnet_1: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_offentlig_utilgjengelig',
  },

  afpOffentligOppgitt_2: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })}`,
    content: 'grunnlag.afp.ingress.ja_offentlig',
  },

  offentligAfpOgUforeKanIkkeBeregnes_3: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_offentlig.ufoeretrygd',
  },

  afpPrivat_4: {
    title: `${intl.formatMessage({ id: 'afp.privat' })}`,
    content: 'grunnlag.afp.ingress.ja_privat',
  },

  afpPrivatIkkeBeregnet_5: {
    title: `${intl.formatMessage({ id: 'afp.nei' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_privat.ufoeretrygd',
  },

  afpIkkeSvart_6: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.nei',
  },

  afpVetIkke_7: {
    title: `${intl.formatMessage({ id: 'afp.vet_ikke' })}`,
    content: 'grunnlag.afp.ingress.vet_ikke',
  },

  afpVetIkkeUforetrygd_8: {
    title: `${intl.formatMessage({ id: 'afp.vet_ikke' })}`,
    content: 'grunnlag.afp.ingress.vet_ikke.ufoeretrygd',
  },

  afpUforetrygd_9: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.overgangskull.ufoeretrygd_eller_ap',
  },

  afpUforetrygdNei_10: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.ufoeretrygd',
  },
  afpPrivatUendret_11: {
    title: `${intl.formatMessage({ id: 'afp.privat' })} (${intl.formatMessage({ id: 'grunnlag.afp.endring' })})`,
    content: 'grunnlag.afp.ingress.ja_privat.endring',
  },
  harAfpOffentlig_12: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })}`,
    content: 'grunnlag.afp.ingress.overgangskull',
  },
})

export const generateAfpContent =
  (intl: IntlShape) => (input: IAfpGrunnlagInput) => {
    const content = afpContentIntl(intl)
    const {
      afpValg,
      samtykkeOffentligAFP,
      afpUtregning,
      foedselsdato,
      loependeVedtak,
      beregningsvalg,
      erApoteker,
    } = input

    const hasUfoeregradGreaterThanZero = loependeVedtak?.ufoeretrygd?.grad > 0
    const has100PercentUfoeretrygd = loependeVedtak?.ufoeretrygd?.grad === 100
    const hasLoependeVedtak = loependeVedtak?.harLoependeVedtak === true
    const isKap19 = isFoedtFoer1963(foedselsdato)
    const hasAfpPrivat = loependeVedtak?.afpPrivat !== undefined
    const hasAfpOffentlig = loependeVedtak?.afpOffentlig !== undefined
    const hasAlderspensjon = loependeVedtak?.alderspensjon !== undefined
    const isAfpValgMissing = afpValg === null || afpValg === undefined

    // Personer født før 1963 eller apotekere behandles likt for AFP-formål
    const isKap19OrApoteker = isKap19 || erApoteker

    // Prioritet 1: Håndter eksisterende AFP-vedtak
    if (hasAfpOffentlig) {
      return content.harAfpOffentlig_12
    }
    if (hasAfpPrivat) {
      return content.afpPrivatUendret_11
    }

    // Prioritet 2: Håndter spesielle regler for apotekere og personer født før 1963
    if (isKap19OrApoteker) {
      // Offentlig AFP med KUN_ALDERSPENSJON
      if (afpValg === 'ja_offentlig' && afpUtregning === 'KUN_ALDERSPENSJON') {
        return content.afpIkkeSvart_6
      }

      // Manglende afpValg med vedtak om alderspensjon eller uføretrygd
      if (
        afpValg === null &&
        (hasAlderspensjon || hasUfoeregradGreaterThanZero)
      ) {
        return content.afpUforetrygd_9
      }
    }

    // Prioritet 3: Håndter tilfeller med uføretrygd
    if (hasUfoeregradGreaterThanZero) {
      // Tilfelle 3.1: Født før 1963 eller apoteker med manglende AFP-valg
      if (isKap19OrApoteker && isAfpValgMissing) {
        return content.afpUforetrygd_9
      }

      // Tilfelle 3.2: Personer som verken er født før 1963 eller er apoteker, med uføretrygd
      if (!isKap19OrApoteker) {
        // Håndter ulike AFP-valg med uføretrygd
        switch (afpValg) {
          case 'ja_offentlig': {
            // Manglende samtykke til AFP-beregning (prioriteres høyest)
            if (samtykkeOffentligAFP === false) {
              return content.offentligIkkeBeregnet_1
            }

            if (beregningsvalg === 'uten_afp' || beregningsvalg === null) {
              return content.offentligAfpOgUforeKanIkkeBeregnes_3
            }

            return content.afpOffentligOppgitt_2
          }

          case 'ja_privat': {
            if (beregningsvalg === 'uten_afp') {
              return content.afpPrivatIkkeBeregnet_5
            }
            return content.afpPrivat_4
          }

          case 'nei': {
            // Spesialtilfelle for gradert vedtak uten 100% uføregrad
            if (!has100PercentUfoeretrygd && !isKap19) {
              return content.afpIkkeSvart_6
            }
            return content.afpUforetrygdNei_10
          }

          case 'vet_ikke': {
            return content.afpVetIkkeUforetrygd_8
          }

          default: {
            // Tilfelle for null/undefined
            if (!isKap19OrApoteker && (hasLoependeVedtak || hasAlderspensjon)) {
              return content.afpUforetrygdNei_10
            }
            return content.afpUforetrygdNei_10
          }
        }
      }
    }

    // Prioritet 4: Håndter AFP-valg basert på forskjellige betingelser (for tilfeller uten uføretrygd)
    switch (afpValg) {
      case 'ja_offentlig': {
        // Tilfelle 4.1: Født 1963 eller senere (og ikke apoteker) med offentlig AFP
        if (!isKap19OrApoteker) {
          // Manglende samtykke til AFP-beregning
          if (samtykkeOffentligAFP === false) {
            return content.offentligIkkeBeregnet_1
          }

          // Samtykket til AFP-beregning
          return content.afpOffentligOppgitt_2
        }

        // Tilfelle 4.2: Født før 1963 eller apoteker med offentlig AFP (isKap19OrApoteker)
        if (afpUtregning === 'KUN_ALDERSPENSJON') {
          return content.afpIkkeSvart_6
        }

        if (hasLoependeVedtak && hasUfoeregradGreaterThanZero) {
          return content.harAfpOffentlig_12
        }

        return content.afpOffentligOppgitt_2
      }

      case 'ja_privat': {
        // Tilfelle 4.3: Privat AFP (standardrespons)
        return content.afpPrivat_4
      }

      case 'nei': {
        // Tilfelle 4.4: Ingen AFP
        return content.afpIkkeSvart_6
      }

      case 'vet_ikke':
      default: {
        // Tilfelle 4.5: "Vet ikke"
        return content.afpVetIkke_7
      }
    }
  }
