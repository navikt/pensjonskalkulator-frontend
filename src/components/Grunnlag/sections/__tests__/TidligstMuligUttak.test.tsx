import { vi } from 'vitest'

import { TidligstMuligUttak } from '../TidligstMuligUttak'
import * as velgUttaksalderUtils from '@/components/VelgUttaksalder/utils'
import { render, screen, userEvent } from '@/test-utils'

describe('TidligstMuligUttak', () => {
  it('rendrer riktig tittel med formatert uttaksalder og tekst', async () => {
    const user = userEvent.setup()
    const formatMock = vi.spyOn(velgUttaksalderUtils, 'formatUttaksalder')
    const { asFragment } = render(
      <TidligstMuligUttak
        uttaksalder={{ aar: 62, maaned: 10, uttaksdato: '2031-11-01' }}
      />
    )
    expect(screen.getByText('Tidligst mulig uttak:')).toBeVisible()
    expect(formatMock).toHaveBeenCalled()
    await user.click(screen.getByRole('button'))
    expect(asFragment()).toMatchSnapshot()
  })
})
