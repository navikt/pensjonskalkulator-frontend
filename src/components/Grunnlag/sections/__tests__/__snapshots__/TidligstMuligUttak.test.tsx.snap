// Vitest Snapshot v1

exports[`TidligstMuligUttak > rendrer riktig tittel med formatert uttaksalder og tekst 1`] = `
<DocumentFragment>
  <div
    class="navds-accordion__item navds-accordion__item--open"
  >
    <button
      aria-expanded="true"
      class="navds-accordion__header"
      data-testid="accordion-header"
      type="button"
    >
      <div
        class="navds-accordion__icon-wrapper"
      >
        <svg
          aria-hidden="true"
          aria-labelledby="title-r0"
          class="navds-accordion__header-chevron"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="title-r0"
          >
            Vis mer
          </title>
          <path
            clip-rule="evenodd"
            d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <span
        class="navds-accordion__header-content navds-heading navds-heading--small"
      >
        Tidligst mulig uttak: 
        <span
          class="_details_c03039"
        >
          62 år og 10 md.
        </span>
      </span>
    </button>
    <div
      class="navds-accordion__content _content_e53ef6 navds-body-long"
    >
      <p
        class="navds-body-long"
      >
        Når du velger hvilken alder du ønsker ta ut pensjon er det alltid måneden etter du fyller år som blir brukt i beregningen. Velger du for eksempel 62 år, betyr det måneden etter du fyller 62 år. 
        <br />
        <br />
        For å starte uttak før 67 år må opptjeningen være høy nok, og det vil derfor ikke være alle som har mulighet til å starte uttak ved 62 år. Tidspunktet som er oppgitt er ett estimat på når du tidligst kan starte uttak av 100 % alderspensjon basert på din pensjonsopptjening og opplysninger du har gitt.
        <br />
        <br />
        Tidspunktet kan bli endret hvis din inntekt endrer seg, og ut fra om du har rett til AFP i privat eller offentlig sektor. Du bør derfor sjekke igjen når det nærmer seg at du ønsker å starte alderspensjon.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
